### ISSUE TITLE
Docker push on private registry hangs

### ISSUE THREAD

**Issue by TrustNoOne** [docker/distribution#469]
Hi,
I set up a private 2.0 registry on azure, used nginx for basic auth and in the end made it all work.
When i started pushing images though, I noticed that almost every time the push hangs indefinitely at some point with something like this:
`[info] 86dfb1389584: Buffering to Disk`
One time I waited for an hour, then ctrl-c'd it and did it again, only to get an error message saying that docker is already pushing that image. I restarted the daemon, pushed it again and it got stuck after a different hash. The third time it succeeded. 

It did that a number of times with different images. I couldn't find anything in the logs (both client and server side) so I debugged the network with iftop: the push is terribly slow (I only have 4 mbits unfortunately) but I can see when it hangs because the connection to the registry disappears completely from the iftop list. 

tl;dr: connection closes, docker deamon hangs indefinitely while pushing.

**Comment by RichardScothern** [comment]
Hi @TrustNoOne thanks for notifying us of this issue.  Could you check a couple of things to help diagnose this problem:
 - What is that status of the socket when the hang occurs (lsof or netstat should give this - it should be something like ESTABLISHED or TIME_WAIT)

 - Could you check the logs around the time the 'Buffering to Disk'.  Specifically there should be something like "rendered layer for ..." to indicate the buffer finished and the push started.  You can get verbose logging by running the daemon as "docker -d -D"

**Comment by TrustNoOne** [comment]
@dmcgowan  @RichardScothern  Here are the logs. I don't see anything suspicios. I hid the company/product name.
It's been  hanging for more than an hour, no cpu/network activity whatsovere.

 This is daemon side:

```
DEBU[483279] Using cached token for ...
DEBU[483279] hostDir: /etc/docker/certs.d/myregistry.myhost.com
DEBU[483352] Pushing layer: b1eaf4accb1ccfb8aa1d0c32ad0e535392eef597e1e53a372d773facb969b6b6
DEBU[483352] [registry] Calling "HEAD" https://myregistry.myhost.com/v2/some/container/blobs/sha256:4dc28a4c114bd64a90d3c82e2ed9b607cdc4b7a10032ebbc15fc215822490fb9
DEBU[483352] https://myregistry.myhost.com/v2/some/container/blobs/sha256:4dc28a4c114bd64a90d3c82e2ed9b607cdc4b7a10032ebbc15fc215822490fb9 -- HEADERS: map[User-Agent:[docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64]]
DEBU[483352] Using cached token for ...
DEBU[483352] hostDir: /etc/docker/certs.d/myregistry.myhost.com
DEBU[483412] rendered layer for b1eaf4accb1ccfb8aa1d0c32ad0e535392eef597e1e53a372d773facb969b6b6 of [181817643] size
DEBU[483412] [registry] Calling "POST" https://myregistry.myhost.com/v2/some/container/blobs/uploads/
DEBU[483412] https://myregistry.myhost.com/v2/some/container/blobs/uploads/ -- HEADERS: map[User-Agent:[docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64]]
DEBU[483412] Using cached token for ...
DEBU[483412] hostDir: /etc/docker/certs.d/myregistry.myhost.com
DEBU[483413] [registry] Calling "PUT" https://myregistry.myhost.com/v2/some/container/blobs/uploads/c9acc202-3d8d-4317-b81e-d92fdd1ee6c3?_state=L90dBG48fSnBfys5AWbCbR-nRY-3IgIMMZKBMPXxfPt7Ik5hbWUiOiJyZWFxdGEvcmVhcXRhLWNvbXBvbmVudC1tYW5hZ2VyIiwiVVVJRCI6ImM5YWNjMjAyLTNkOGQtNDMxNy1iODFlLWQ5MmZkZDFlZTZjMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNS0wNS0wNVQwNzo0OTo0Mi4zNTg4MVoifQ%3D%3D
DEBU[483413] https://myregistry.myhost.com/v2/some/container/blobs/uploads/c9acc202-3d8d-4317-b81e-d92fdd1ee6c3?_state=L90dBG48fSnBfys5AWbCbR-nRY-3IgIMMZKBMPXxfPt7Ik5hbWUiOiJyZWFxdGEvcmVhcXRhLWNvbXBvbmVudC1tYW5hZ2VyIiwiVVVJRCI6ImM5YWNjMjAyLTNkOGQtNDMxNy1iODFlLWQ5MmZkZDFlZTZjMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNS0wNS0wNVQwNzo0OTo0Mi4zNTg4MVoifQ%3D%3D -- HEADERS: map[User-Agent:[docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64]]
DEBU[483413] Using cached token for ...
DEBU[483413] hostDir: /etc/docker/certs.d/myregistry.myhost.com
```

This is registry side:
```
time="2015-05-05T08:00:58.0002924Z" level=info msg=Base.WriteStream trace.duration=5m21.7298247s trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).WriteStream" trace.id=1823bd63-4e7a-4a21-8009-a0632939ad0b trace.line=95
time="2015-05-05T08:00:58.0006111Z" level=debug msg="(*layerWriter).Finish" environment=prod http.request.host=myregistry.myhost.com http.request.id=56b9c6cb-8d6b-4aa6-ad05-a79b8ef379c0 http.request.method=PUT http.request.remoteaddr=212.56.148.178 http.request.uri="/v2/some/container/blobs/uploads/c9acc202-3d8d-4317-b81e-d92fdd1ee6c3?_state=L90dBG48fSnBfys5AWbCbR-nRY-3IgIMMZKBMPXxfPt7Ik5hbWUiOiJyZWFxdGEvcmVhcXRhLWNvbXBvbmVudC1tYW5hZ2VyIiwiVVVJRCI6ImM5YWNjMjAyLTNkOGQtNDMxNy1iODFlLWQ5MmZkZDFlZTZjMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNS0wNS0wNVQwNzo0OTo0Mi4zNTg4MVoifQ%3D%3D&digest=sha256%3A4dc28a4c114bd64a90d3c82e2ed9b607cdc4b7a10032ebbc15fc215822490fb9" http.request.useragent="docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64" instance.id=ed085fce-28d0-4b18-b85d-a51f85f95bd4 service=registry vars.name="some/container" vars.uuid=c9acc202-3d8d-4317-b81e-d92fdd1ee6c3 version=v2.0.0.m
time="2015-05-05T08:00:58.1686527Z" level=info msg=Base.Stat trace.duration=167.7249ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=2bf900e0-c149-4461-b8b0-0f5f5e4e06cb trace.line=111
time="2015-05-05T08:00:58.3703744Z" level=info msg=Base.Stat trace.duration=201.4775ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=c4f607bf-1322-4997-8269-bac9fe6b38ca trace.line=111
time="2015-05-05T08:00:58.5157366Z" level=info msg=Base.Move trace.duration=145.1079ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Move" trace.id=954dbd11-8e5d-43c4-ba66-0f9a8323e5ae trace.line=135
time="2015-05-05T08:00:58.870929Z" level=info msg=Base.PutContent trace.duration=354.9186ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).PutContent" trace.id=07d08022-0c5a-423a-b9f7-68f8aa84fc58 trace.line=67
time="2015-05-05T08:00:59.1511369Z" level=info msg=Base.Move trace.duration=279.9499ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Delete" trace.id=f11a9d8d-9b78-4a0e-9a36-061da4342ed0 trace.line=149
time="2015-05-05T08:00:59.1514745Z" level=debug msg="(*layerStore).Fetch" environment=prod http.request.host=myregistry.myhost.com http.request.id=56b9c6cb-8d6b-4aa6-ad05-a79b8ef379c0 http.request.method=PUT http.request.remoteaddr=212.56.148.178 http.request.uri="/v2/some/container/blobs/uploads/c9acc202-3d8d-4317-b81e-d92fdd1ee6c3?_state=L90dBG48fSnBfys5AWbCbR-nRY-3IgIMMZKBMPXxfPt7Ik5hbWUiOiJyZWFxdGEvcmVhcXRhLWNvbXBvbmVudC1tYW5hZ2VyIiwiVVVJRCI6ImM5YWNjMjAyLTNkOGQtNDMxNy1iODFlLWQ5MmZkZDFlZTZjMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNS0wNS0wNVQwNzo0OTo0Mi4zNTg4MVoifQ%3D%3D&digest=sha256%3A4dc28a4c114bd64a90d3c82e2ed9b607cdc4b7a10032ebbc15fc215822490fb9" http.request.useragent="docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64" instance.id=ed085fce-28d0-4b18-b85d-a51f85f95bd4 service=registry vars.name="some/container" vars.uuid=c9acc202-3d8d-4317-b81e-d92fdd1ee6c3 version=v2.0.0.m
time="2015-05-05T08:00:59.2669982Z" level=info msg=Base.GetContent trace.duration=115.3228ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=d0fa40a8-f30d-4e4a-a217-8764ab7da880 trace.line=55
time="2015-05-05T08:00:59.284431Z" level=info msg=Base.Stat trace.duration=17.1573ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=5a01272e-afbb-4391-a2c9-f1d106ebdaa8 trace.line=111
time="2015-05-05T08:00:59.2948302Z" level=info msg=Base.Stat trace.duration=10.2156ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=a59f3ddf-ee31-449f-869f-962d1cc20db5 trace.line=111
time="2015-05-05T08:00:59.3047749Z" level=info msg="response completed" environment=prod http.request.host=myregistry.myhost.com http.request.id=56b9c6cb-8d6b-4aa6-ad05-a79b8ef379c0 http.request.method=PUT http.request.remoteaddr=212.56.148.178 http.request.uri="/v2/some/container/blobs/uploads/c9acc202-3d8d-4317-b81e-d92fdd1ee6c3?_state=L90dBG48fSnBfys5AWbCbR-nRY-3IgIMMZKBMPXxfPt7Ik5hbWUiOiJyZWFxdGEvcmVhcXRhLWNvbXBvbmVudC1tYW5hZ2VyIiwiVVVJRCI6ImM5YWNjMjAyLTNkOGQtNDMxNy1iODFlLWQ5MmZkZDFlZTZjMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNS0wNS0wNVQwNzo0OTo0Mi4zNTg4MVoifQ%3D%3D&digest=sha256%3A4dc28a4c114bd64a90d3c82e2ed9b607cdc4b7a10032ebbc15fc215822490fb9" http.request.useragent="docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64" http.response.duration=5m24.0917414s http.response.written=0 instance.id=ed085fce-28d0-4b18-b85d-a51f85f95bd4 service=registry version=v2.0.0.m
172.17.0.4 - - [05/May/2015:07:55:35 +0000] "PUT /v2/some/container/blobs/uploads/c9acc202-3d8d-4317-b81e-d92fdd1ee6c3?_state=L90dBG48fSnBfys5AWbCbR-nRY-3IgIMMZKBMPXxfPt7Ik5hbWUiOiJyZWFxdGEvcmVhcXRhLWNvbXBvbmVudC1tYW5hZ2VyIiwiVVVJRCI6ImM5YWNjMjAyLTNkOGQtNDMxNy1iODFlLWQ5MmZkZDFlZTZjMyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNS0wNS0wNVQwNzo0OTo0Mi4zNTg4MVoifQ%3D%3D&digest=sha256%3A4dc28a4c114bd64a90d3c82e2ed9b607cdc4b7a10032ebbc15fc215822490fb9 HTTP/1.0" 201 0 "" "docker/1.6.0 go/go1.4.2 git-commit/4749651 kernel/3.16.0-34-generic os/linux arch/amd64"
```

This is the client log:
```
[info] 10fdd99f8cd4: Buffering to Disk
[info] 10fdd99f8cd4: Image successfully pushed
[info] 10fdd99f8cd4: Image already exists
[info] dab408d4d349: Buffering to Disk
[info] dab408d4d349: Image successfully pushed
[info] 0dd1b72f11a9: Buffering to Disk
[info] 0dd1b72f11a9: Image successfully pushed
[info] 9b973c7b53b0: Buffering to Disk
[info] 9b973c7b53b0: Image successfully pushed
[info] 0845cee9ffcb: Buffering to Disk
[info] 0845cee9ffcb: Image successfully pushed
[info] 73abf6e6861a: Buffering to Disk
[info] 73abf6e6861a: Image successfully pushed
[info] a331aad3b6c0: Buffering to Disk
[info] a331aad3b6c0: Image successfully pushed
[info] b1eaf4accb1c: Buffering to Disk
```

**Comment by TrustNoOne** [comment]
Ok, it's stuck again, same logs as before, socket ESTABLISHED
```tcp        0      0 192.168.30.14:35452     x.y.z:443       ESTABLISHED```

maybe something related to nginx/timeouts?

**Comment by RichardScothern** [comment]
This looks more like an issue with the docker daemon interacting with nginx.   In that case, this issue belongs in the docker project.

For a more reproducible test scenario, you could switch from azure to filesystem and rm -rf the directory to reproduce uploading the layer data.

**Comment by ledjon** [comment]
I'm having this same issue and not doing any sort of nginx proxy.  I am just running the registry and hitting it directly.

So is this an issue that there is a workaround for?

**Comment by TrustNoOne** [comment]
It still does that to me. I cannot switch from azure to filesystem because it's a registry with production releases and we're a small startup so unfortunately I don't have time to do other tests :\

**Comment by stevvooe** [comment]
From the registry logs, it looks like the request was actually completed. Likely, the proxy dropped the connection and the daemon never received the response. This is likely a daemon bug if it is not timing out properly.

I suspect this problem is compounded by a large image. While registry v2 protocol can handle very large images, the daemon may struggle pushing very large layers. Breaking these layers up almost always works better.

You may want to investigate the proxy settings to see if the timeout values are set appropriately. With a 4MBit connection fully filled, you are looking at around 35 minutes to push a gigabyte image in ideal conditions. Given other factors, such as azure round trips (which look decent from the logs), this can compound.

The other possibility is the following: https://github.com/docker/docker/issues/13250#issuecomment-102547028

Check out some of the debugging described there. Again, smaller layer sizes will help to avoid these problems.

**Comment by MrJoy** [comment]
I'm seeing a similar issue with registry 2.0.1 and docker 1.7.0 except the layers involved are not large.  I've seen it hang on layers as small as tens of kilobytes.  In my case though, I'm storing to S3.  Note that the issue does *not* happen for me if I use local storage.

**Comment by ledjon** [comment]
Yes we are at the same point -- we have switched from s3 to local storage to get around the problem (works 100% of the time on local storage).  Obviously we lose a lot of flexibility this way, but at least we aren't getting failures 30% of the time on uploads.

**Comment by MrJoy** [comment]
I'm getting failures 100% of the time, but switching to local storage is just not an acceptable option.  I'll likely punt and roll back to registry 1.x until this issue is fixed.  Short form:  Our devops infrastructure is not yet fleshed out enough to handle persistent local storage gracefully and we have some rather higher priority items to finish up first (mostly around establishing security zones, fully isolating employees within them, etc...).

**Comment by stevvooe** [comment]
We've put together docker/docker#14018 to address some of the push performance problems in the client. While I don't think this issue is a docker daemon bug completely, reading through that might provide a little background.

The primary problem covered in this issue (as reported by @TrustNoOne) is slow push performance when running a registry remote to a storage backend over a slow or narrow (bandwidth) link (s3, azure, etc.). I'll change the title accordingly. This can be affected by image size but is not dependent on it. The primary cause is slow bandwidth to the storage backend which can be compounded by latency.

@MrJoy It might help if you could provide a little more information about your deployment. While you may be getting the same behavior as @TrustNoOne, it is hard to say if it has the same cause. Please provide some information about where you are deploying the registry (in AWS or elsewhere), what kind of network connection you have with the backend and we can diagnose your problem. Because this is a problem of setup, rather than a specific problem, waiting for this get "fixed" will not likely be successful.

One thing you could try with the s3 driver is to decrease the [`chunksize`](https://github.com/docker/distribution/blob/master/docs/configuration.md#s3). For your bandwidth allocation, 800KB or less would be optimal but the minimum value that can be set is 5242880.

**Comment by wpp** [comment]
Hi @stevvooe since you mentioned

> the daemon may struggle pushing very large layers. Breaking these layers up almost always works better.

What would qualify as a *big layer*?

**Comment by stevvooe** [comment]
@wpp As a rule of thumb, I would say >500MB. This is going to vary based on host CPU, disk and network. The slower the pipe, the smaller the bound for what constitutes a "big image".

Layers are basically the update granularity. Reducing the size of layers will reduce the size of a differential push.

**Comment by y0n1** [comment]
Same happened to me on Windows, I worked it around by restarting the virtual machine (`docker-machine restart`).

**Comment by openjack** [comment]
I'm using Mac Beta, I've been dealing with this problem for a few hours.  Restarting the docker daemon worked for me as well.

**Comment by stevvooe** [comment]
@y0n1 @openjack I'd recommend reporting through docker mac/windows channels. This is unlikely to be the same issue.

**Comment by monikakatiyar16** [comment]
Attaching reference to the issue resolution for me when using nginx as reverse proxy
https://github.com/docker/docker/issues/2292#issuecomment-267664408

**Comment by raam86** [comment]
Still reproducible, when pushing to AWS ECS, restarting the docker-machine worked.

**Comment by stevvooe** [comment]
Reading back through here, this really sounds like a docker issue, rather than a registry issue. I'd recommend those still looking for help to file an issue in docker/docker. Either way, I think we can better address the problems there.

**Comment by sblackstone** [comment]
Just had this occur, would hang while pushing a layer to a private registry and finally die with a socket error.

First, I tried clearing out the registry's data folder and restarting it thinking it could be some sort of corruption but after re-pushing I still had the same error (albeit on a different file this time)..

I then restarted Docker on my mac and it was fixed.

Docker Version = 18.03.0-ce-mac60 (23751)

**Comment by twiden** [comment]
I had the same issue. Running a private registry on a Linux VM and pushing image from the host (OSX 10.13.4). The push would hang on a 24Mb layer every time.

Restarted docker (18.03.1-ce, build 9ee9f40) and the issue was resolved.

**Comment by kethansrinivas9** [comment]
I had the same issue in windows with docker in minikube environment, it worked fine after restarting the minikube virtual machine.

**Comment by Darkeye9** [comment]
5 Years later, the docker daemon solutions works perfectly!

**Comment by triwats** [comment]
Hit this today, restarting the docker agent worked for me, try that first before investigating further.

**Comment by james-mchugh** [comment]
I still hit this issue from time to time, and restarting the Docker daemon does not help.

**Comment by adlion** [comment]
I am hopeless.
Since last hours I am pushing one image in gitlab registry and it freeze on the second layer. Always on pushing.
I changed the daemon to allow one connection, and is working one time only and the next time I push it freezes itself.
Tried to uninstall, restart, push in another registry, nothing help.

**Comment by adlion** [comment]
I had the same version.
Tested on two local machines and same issue.
Than tested on a server on DO and no issue at all.
So I installed the version I had on DO and everything is going fine so far.
The version I had the error was `19.03.13, build 4484c46d9d` and the version on DO is `19.03.9, build 9d988398e7`.

**Comment by zhangguanzhang** [comment]
same issue with `19.03.14`

**Comment by bensialih** [comment]
on ubuntu I restarted docker service and that fixed the hanging problem
`sudo service docker restart`
